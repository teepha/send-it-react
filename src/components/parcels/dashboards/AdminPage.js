import React from "react";
import Spinner from "react-md-spinner";
import { withRouter } from "react-router-dom";
import { connect } from "react-redux";
import { toast } from "react-toastify";
import {
  getAllParcels,
  updateParcelStatus,
  updateParcelLocation
} from "../../../actions/parcelsActions";
import ViewModal from "../../common/modals/ViewModal";
import LocationModal from "../../common/modals/LocationModal";

export class AdminProfile extends React.Component {
  state = {
    parcels: [],
    noParcelsErrMsg: "",
    search: "",
    parcelsCopy: [],
    viewModalIsOpen: false,
    parcelToView: {},
    locationModalIsOpen: false,
    parcelToUpdate: {},
    parcelToUpdateId: "",
    newLocation: "",
    presentLocation: ""
  };

  componentDidMount() {
    this.props.getAllParcels();
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (nextProps.error && this.props.error !== nextProps.error) {
      const errorString = nextProps.error;
      this.setState({ noParcelsErrMsg: errorString });
    } else if (this.props.parcels !== nextProps.parcels) {
      this.setState({
        parcels: nextProps.parcels,
        parcelsCopy: nextProps.parcels
      });
    }
    return true;
  }

  handleInputChange = e => {
    const { parcelsCopy } = this.state;
    const value = e.target.value.trim();
    const filteredParcels = parcelsCopy.filter(parcel =>
      parcel.user_id.toString().includes(value)
    );
    this.setState({ parcels: filteredParcels, search: value });
  };

  handleStatusChange = e => {
    const { id, value } = e.target;
    this.props.updateParcelStatus(id, value);
    toast.success("Parcel Status Updated Successfully!");
  };

  handleOpenViewModal = event => {
    const id = parseInt(event.target.id, 10);
    const findParcel = this.state.parcels.find(parcel => parcel.id === id);
    this.setState({ viewModalIsOpen: true, parcelToView: findParcel });
  };

  handleCloseViewModal = () => {
    this.setState({ viewModalIsOpen: false });
  };

  handleOpenLocationModal = event => {
    const id = parseInt(event.target.id, 10);
    const findParcel = this.state.parcels.find(parcel => parcel.id === id);
    this.setState({
      locationModalIsOpen: true,
      parcelToUpdate: findParcel,
      parcelToUpdateId: id
    });
  };

  handleLocationInputChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  handleLocationUpdate = e => {
    e.preventDefault();
    const { parcelToUpdateId, newLocation } = this.state;
    this.props.updateParcelLocation(parcelToUpdateId, newLocation);
    this.setState({ locationModalIsOpen: false });
    toast.success("Parcel Location Successfully Updated!");
  };

  handleCloseLocationModal = () => {
    this.setState({ locationModalIsOpen: false });
  };

  render() {
    const {
      noParcelsErrMsg,
      parcels,
      parcelToView,
      parcelToUpdate
    } = this.state;
    const { loading } = this.props;

    return loading ? (
      <div
        style={{
          display: "flex",
          justifyContent: "center",
          marginTop: "15%",
          marginBottom: "15%"
        }}
      >
        <Spinner singleColor="blue" size={70} />
      </div>
    ) : (
      <div>
        {noParcelsErrMsg ? (
          <h1 id="error-msg">{noParcelsErrMsg}</h1>
        ) : (
          <div className="main-content">
            <div className="admin-profile-wrapper">
              <div className="order-count">
                <div>
                  Ready for Pickup:{" "}
                  <span id="pickup-status">
                    {
                      parcels.filter(
                        parcel => parcel.status === "ready_for_pickup"
                      ).length
                    }
                  </span>
                </div>
                <div className="status-count">
                  In-Transit:{" "}
                  <span id="transit-status">
                    {
                      parcels.filter(parcel => parcel.status === "in_transit")
                        .length
                    }
                  </span>
                </div>
                <div className="status-count">
                  Delivered:{" "}
                  <span id="deliver-status">
                    {
                      parcels.filter(parcel => parcel.status === "delivered")
                        .length
                    }
                  </span>
                </div>
                <div className="status-count">
                  Cancelled:{" "}
                  <span id="cancel-status">
                    {
                      parcels.filter(parcel => parcel.status === "cancelled")
                        .length
                    }
                  </span>
                </div>
              </div>

              <div className="table-top">
                <span className="table-title">All Orders</span>
                <div className="search-bar">
                  <input
                    id="search-user-id"
                    type="text"
                    placeholder="Enter User ID..."
                    name="search"
                    value={this.state.search}
                    onChange={this.handleInputChange}
                  />
                </div>
              </div>

              <table className="orders">
                <thead>
                  <tr>
                    <th scope="row">ID</th>
                    <th scope="row" className="remove-first">
                      User ID
                    </th>
                    <th scope="row" className="remove-second">
                      Date
                    </th>
                    <th scope="row" className="remove-first">
                      Pick Up Location
                    </th>
                    <th scope="row">Destination</th>
                    <th scope="row" className="remove-second">
                      Recipient Name
                    </th>
                    <th scope="row">Status</th>
                    <th scope="row">Present Location</th>
                    <th scope="row" />
                    <th scope="row" />
                  </tr>
                </thead>
                <tbody className="orders-data">
                  {parcels.map((parcel, i) => (
                    <tr key={i}>
                      <td>{parcel.id}</td>
                      <td className="remove-first">{parcel.user_id}</td>
                      <td className="remove-second">
                        {parcel.date.slice(0, 10)}
                      </td>
                      <td className="remove-first">{parcel.pickup_location}</td>
                      <td>{parcel.destination}</td>
                      <td className="remove-second">{parcel.recipient_name}</td>
                      <td>
                        <select
                          id={parcel.id}
                          disabled={
                            parcel.status === "cancelled" ||
                            parcel.status === "delivered"
                          }
                          defaultValue={parcel.status}
                          name="status"
                          className="status"
                          onChange={this.handleStatusChange}
                        >
                          <option disabled value="cancelled">
                            Cancelled
                          </option>
                          <option value="ready_for_pickup">
                            Ready For Pickup
                          </option>
                          <option value="in_transit">In-Transit</option>
                          <option value="delivered">Delivered</option>
                        </select>
                      </td>
                      <td>{parcel.present_location}</td>
                      <td className="view">
                        <i
                          id={parcel.id}
                          className="far fa-eye"
                          onClick={this.handleOpenViewModal}
                        />
                      </td>
                      <td className="edit">
                        {parcel.status !== "cancelled" &&
                        parcel.status !== "delivered" ? (
                          <i
                            id={parcel.id}
                            className="far fa-edit"
                            onClick={this.handleOpenLocationModal}
                          />
                        ) : (
                          ""
                        )}
                      </td>
                    </tr>
                  ))}
                </tbody>
              </table>

              {/* View Modal */}
              <ViewModal
                isOpen={this.state.viewModalIsOpen}
                onRequestClose={this.handleCloseViewModal}
                parcel={parcelToView}
                handleClick={this.handleCloseViewModal}
              />

              {/* Update Location Modal */}
              <LocationModal
                isOpen={this.state.locationModalIsOpen}
                onRequestClose={this.handleCloseLocationModal}
                parcel={parcelToUpdate}
                handleClick={this.handleCloseLocationModal}
                handleSubmit={this.handleLocationUpdate}
                handleChange={this.handleLocationInputChange}
              />
            </div>
          </div>
        )}
      </div>
    );
  }
}

const mapStateToProps = ({ parcels }) => {
  return {
    loading: parcels.isLoading,
    parcels: parcels.data,
    error: parcels.error
  };
};

const mapDispatchToProps = () => ({
  getAllParcels,
  updateParcelStatus,
  updateParcelLocation
});

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps()
  )(AdminProfile)
);
