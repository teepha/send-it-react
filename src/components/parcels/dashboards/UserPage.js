import React from "react";
import Spinner from "react-md-spinner";
import { Link, withRouter } from "react-router-dom";
import { connect } from "react-redux";
import { toast } from "react-toastify";
import {
  getUserParcels,
  getSingleParcel,
  cancelParcelOrder
} from "../../../actions/parcelsActions";
import { capitalizeStatus } from "../../../utils";
import CancelModal from "../../common/Modals/CancelModal";
import ViewModal from "../../common/Modals/ViewModal";

class UserProfile extends React.Component {
  state = {
    parcels: [],
    noParcelsErrMsg: "",
    cancelModalIsOpen: "",
    search: "",
    parcelsCopy: [],
    viewModalIsOpen: false,
    parcelToView: {},
    cancelModalIsOpen: false,
    parcelToCancelId: ""
  };

  componentDidMount() {
    const { user } = this.props;
    const userId = user.id;
    this.props.getUserParcels(userId);
  }

  shouldComponentUpdate(nextProps, nextState) {
    if (this.props.error !== nextProps.error) {
      const errorString = nextProps.error;
      this.setState({ noParcelsErrMsg: errorString });
    } else if (this.props.parcels !== nextProps.parcels) {
      this.setState({
        parcels: nextProps.parcels,
        parcelsCopy: nextProps.parcels
      });
    }
    return true;
  }

  handleInputChange = e => {
    const { parcelsCopy } = this.state;
    const value = e.target.value.trim().toLowerCase();
    const filteredParcels = parcelsCopy.filter(parcel =>
      parcel.recipient_name.toLowerCase().includes(value)
    );
    this.setState({ parcels: filteredParcels, search: value });
  };

  handleOpenViewModal = event => {
    const id = parseInt(event.target.id, 10);
    const findParcel = this.state.parcels.find(parcel => parcel.id === id);
    this.setState({ viewModalIsOpen: true, parcelToView: findParcel });
  };

  handleCloseViewModal = () => {
    this.setState({ viewModalIsOpen: false });
  };

  handleCancelOrder = e => {
    e.preventDefault();
    this.props.cancelParcelOrder(this.state.parcelToCancelId);
    this.setState({ cancelModalIsOpen: false });
    toast.success("Parcel Order Cancelled Successfully!");
  };

  handleOpenCancelModal = event => {
    const id = parseInt(event.target.id, 10);
    this.setState({
      cancelModalIsOpen: true,
      parcelToCancelId: id
    });
  };

  handleCloseCancelModal = () => {
    this.setState({ cancelModalIsOpen: false });
  };

  render() {
    const { noParcelsErrMsg, parcels, parcelToView } = this.state;
    const { fetching } = this.props;
    return fetching ? (
      <div
        style={{
          display: "flex",
          justifyContent: "center",
          marginTop: "15%",
          marginBottom: "15%"
        }}
      >
        <Spinner singleColor="blue" size={70} />
      </div>
    ) : (
      <div>
        <div className="main-content">
          <div className="user-profile-wrapper">
            <div className="order-count">
              <div>
                Ready for Pickup:
                <span id="pickup-status">
                  {
                    parcels.filter(
                      parcel => parcel.status === "ready_for_pickup"
                    ).length
                  }
                </span>
              </div>
              <div className="status-count">
                In-Transit:
                <span id="transit-status">
                  {
                    parcels.filter(parcel => parcel.status === "in_transit")
                      .length
                  }
                </span>
              </div>
              <div className="status-count">
                Delivered:
                <span id="deliver-status">
                  {
                    parcels.filter(parcel => parcel.status === "delivered")
                      .length
                  }
                </span>
              </div>
              <div className="status-count">
                Cancelled:
                <span id="cancel-status">
                  {
                    parcels.filter(parcel => parcel.status === "cancelled")
                      .length
                  }
                </span>
              </div>
            </div>

            <div className="create-order-btn">
              <Link to="/create-order">Create New Order</Link>
            </div>
            {noParcelsErrMsg ? (
              <h1 id="error-msg">{noParcelsErrMsg}</h1>
            ) : (
              <div>
                <div className="table-top">
                  <span className="table-title">My Orders</span>
                  <div className="search-bar">
                    <input
                      id="search-name"
                      type="text"
                      placeholder="Enter Recipient Name..."
                      name="search"
                      value={this.state.search}
                      onChange={this.handleInputChange}
                    />
                  </div>
                </div>

                <table className="orders">
                  <thead>
                    <tr>
                      <th scope="row">ID</th>
                      <th scope="row" className="remove-second">
                        Date
                      </th>
                      <th scope="row" className="remove-first">
                        Pick Up Location
                      </th>
                      <th scope="row" className="remove-second">
                        Destination
                      </th>
                      <th scope="row">Recipient Name</th>
                      <th scope="row">Status</th>
                      <th scope="row" />
                      <th scope="row" />
                      <th scope="row" />
                    </tr>
                  </thead>
                  <tbody className="orders-data">
                    {parcels.map((parcel, i) => (
                      <tr key={i}>
                        <td>{parcel.id}</td>
                        <td className="remove-second">
                          {parcel.date.slice(0, 10)}
                        </td>
                        <td className="remove-first">
                          {parcel.pickup_location}
                        </td>
                        <td className="remove-second">{parcel.destination}</td>
                        <td>{parcel.recipient_name}</td>
                        <td>
                          {capitalizeStatus(parcel.status.replace(/_/g, " "))}
                        </td>
                        <td className="view">
                          <i
                            id={parcel.id}
                            className="far fa-eye"
                            onClick={this.handleOpenViewModal}
                          />
                        </td>
                        <td>
                          {parcel.status !== "cancelled" &&
                          parcel.status !== "delivered" ? (
                            <Link to={`/parcels/${parcel.id}`}>
                              <i id={parcel.id} className="far fa-edit" />
                            </Link>
                          ) : (
                            ""
                          )}
                        </td>
                        <td className="cancel">
                          {parcel.status !== "cancelled" &&
                          parcel.status !== "delivered" ? (
                            <i
                              id={parcel.id}
                              className="fas fa-times"
                              onClick={this.handleOpenCancelModal}
                            />
                          ) : (
                            ""
                          )}
                        </td>
                      </tr>
                    ))}
                  </tbody>
                </table>

                {/* View Modal */}
                <ViewModal
                  isOpen={this.state.viewModalIsOpen}
                  onRequestClose={this.handleCloseViewModal}
                  parcel={parcelToView}
                  handleClick={this.handleCloseViewModal}
                />

                {/* Cancel Modal */}
                <CancelModal
                  isOpen={this.state.cancelModalIsOpen}
                  onRequestClose={this.handleCloseCancelModal}
                  parcel={parcelToView}
                  handleClick={this.handleCloseCancelModal}
                  handleSubmit={this.handleCancelOrder}
                />
              </div>
            )}
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = ({ user, parcels }) => {
  return {
    fetching: parcels.isFetching,
    user: user.userData,
    parcels: parcels.data,
    error: parcels.error
  };
};

const mapDispatchToProps = () => ({
  getUserParcels,
  getSingleParcel,
  cancelParcelOrder
});

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps()
  )(UserProfile)
);
