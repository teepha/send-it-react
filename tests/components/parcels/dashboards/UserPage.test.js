import React from "react";
import { shallow } from "enzyme";
import { UserProfile } from "../../../../src/components/parcels/dashboards/UserPage";
import { parcelData } from "../../../mockData/testData";

describe("LoginPage Component Test", () => {
  const setUp = () => {
    const props = {
      getUserParcels: jest.fn(() => Promise.resolve()),
      getSingleParcel: jest.fn(() => Promise.resolve()),
      cancelParcelOrder: jest.fn(() => Promise.resolve()),
      loading: false,
      user: {},
      parcels: [],
      error: "",
      history: {
        push: jest.fn()
      }
    };
    const state = {
      parcels: [],
      noParcelsErrMsg: "",
      cancelModalIsOpen: "",
      search: "",
      parcelsCopy: [],
      viewModalIsOpen: false,
      parcelToView: {},
      cancelModalIsOpen: false,
      parcelToCancelId: ""
    };
    return {
      shallowWrapper: shallow(<UserProfile {...props} />),
      state,
      props
    };
  };
  const nextProps = {
    parcels: [],
    error: ""
  };
  const event = {
    target: {
      value: ""
    },
    preventDefault: jest.fn()
  };

  it("render without crashing", () => {
    const { shallowWrapper } = setUp();
    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it("should display <Spinner /> component when `isLoading` is set to true", () => {
    const { shallowWrapper, props } = setUp();
    shallowWrapper.setProps({
      ...props,
      loading: true
    });
    expect(shallowWrapper.find("MDSpinner")).toHaveLength(1);
  });

  it("invokes handleInputChange method", () => {
    const { shallowWrapper, state } = setUp();
    const handleInputChangeSpy = jest.spyOn(
      shallowWrapper.instance(),
      "handleInputChange"
    );
    shallowWrapper.instance().handleInputChange(event);
    expect(handleInputChangeSpy).toHaveBeenCalled();

    shallowWrapper.setState({
      ...state,
      parcelsCopy: parcelData.parcels
    });
    shallowWrapper.instance().handleInputChange(event);
    expect(handleInputChangeSpy).toHaveBeenCalled();
  });

  it("invokes handleOpenViewModal method", () => {
    const { shallowWrapper, state } = setUp();
    const handleOpenViewModalSpy = jest.spyOn(
      shallowWrapper.instance(),
      "handleOpenViewModal"
    );
    shallowWrapper.instance().handleOpenViewModal(event);
    expect(handleOpenViewModalSpy).toHaveBeenCalled();

    shallowWrapper.setState({
      ...state,
      parcels: parcelData.parcels
    });
    shallowWrapper.instance().handleOpenViewModal(event);
    expect(handleOpenViewModalSpy).toHaveBeenCalled();
  });

  it("invokes handleCloseViewModal method", () => {
    const { shallowWrapper } = setUp();
    const handleCloseViewModalSpy = jest.spyOn(
      shallowWrapper.instance(),
      "handleCloseViewModal"
    );
    shallowWrapper.instance().handleCloseViewModal();
    expect(handleCloseViewModalSpy).toHaveBeenCalled();
  });

  it("invokes handleCancelOrder method", () => {
    const { shallowWrapper, state } = setUp();
    const handleCancelOrderSpy = jest.spyOn(
      shallowWrapper.instance(),
      "handleCancelOrder"
    );
    shallowWrapper.instance().handleCancelOrder(event);
    expect(handleCancelOrderSpy).toHaveBeenCalled();

    shallowWrapper.setState({
      ...state,
      parcels: parcelData.parcels
    });
    shallowWrapper.instance().handleCancelOrder(event);
    expect(handleCancelOrderSpy).toHaveBeenCalled();
  });

  it("invokes handleCloseCancelModal method", () => {
    const { shallowWrapper } = setUp();
    const handleCloseCancelModalSpy = jest.spyOn(
      shallowWrapper.instance(),
      "handleCloseCancelModal"
    );
    shallowWrapper.instance().handleCloseCancelModal();
    expect(handleCloseCancelModalSpy).toHaveBeenCalled();
  });

  it("invokes handleOpenCancelModal method", () => {
    const { shallowWrapper, state } = setUp();
    const handleOpenCancelModalSpy = jest.spyOn(
      shallowWrapper.instance(),
      "handleOpenCancelModal"
    );
    shallowWrapper.instance().handleOpenCancelModal(event);
    expect(handleOpenCancelModalSpy).toHaveBeenCalled();
  });

  it("invokes shouldComponentUpdate method", () => {
    const { shallowWrapper } = setUp();
    const shouldComponentUpdateSpy = jest.spyOn(
      shallowWrapper.instance(),
      "shouldComponentUpdate"
    );

    shallowWrapper.setProps({
      ...nextProps,
      error: parcelData.parcelsErrorResponse
    });
    shallowWrapper.instance().shouldComponentUpdate(nextProps);
    expect(shouldComponentUpdateSpy).toHaveBeenCalled();

    shallowWrapper.setProps({
      ...nextProps,
      parcels: parcelData.parcels
    });
    shallowWrapper.instance().shouldComponentUpdate(nextProps);
    expect(shouldComponentUpdateSpy).toHaveBeenCalled();
  });
});
