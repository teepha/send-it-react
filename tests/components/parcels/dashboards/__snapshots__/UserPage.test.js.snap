// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage Component Test render without crashing 1`] = `
<div>
  <div
    className="main-content"
  >
    <div
      className="user-profile-wrapper"
    >
      <div
        className="order-count"
      >
        <div>
          Ready for Pickup:
          <span
            id="pickup-status"
          >
            0
          </span>
        </div>
        <div
          className="status-count"
        >
          In-Transit:
          <span
            id="transit-status"
          >
            0
          </span>
        </div>
        <div
          className="status-count"
        >
          Delivered:
          <span
            id="deliver-status"
          >
            0
          </span>
        </div>
        <div
          className="status-count"
        >
          Cancelled:
          <span
            id="cancel-status"
          >
            0
          </span>
        </div>
      </div>
      <div
        className="create-order-btn"
      >
        <Link
          replace={false}
          to="/create-order"
        >
          Create New Order
        </Link>
      </div>
      <div>
        <div
          className="table-top"
        >
          <span
            className="table-title"
          >
            My Orders
          </span>
          <div
            className="search-bar"
          >
            <input
              id="search-name"
              name="search"
              onChange={[Function]}
              placeholder="Enter Recipient Name..."
              type="text"
              value=""
            />
          </div>
        </div>
        <table
          className="orders"
        >
          <thead>
            <tr>
              <th
                scope="row"
              >
                ID
              </th>
              <th
                className="remove-second"
                scope="row"
              >
                Date
              </th>
              <th
                className="remove-first"
                scope="row"
              >
                Pick Up Location
              </th>
              <th
                className="remove-second"
                scope="row"
              >
                Destination
              </th>
              <th
                scope="row"
              >
                Recipient Name
              </th>
              <th
                scope="row"
              >
                Status
              </th>
              <th
                scope="row"
              />
              <th
                scope="row"
              />
              <th
                scope="row"
              />
            </tr>
          </thead>
          <tbody
            className="orders-data"
          />
        </table>
        <ViewModal
          handleClick={[Function]}
          isOpen={false}
          onRequestClose={[Function]}
          parcel={Object {}}
        />
        <CancelModal
          handleClick={[Function]}
          handleSubmit={[Function]}
          isOpen={false}
          onRequestClose={[Function]}
          parcel={Object {}}
        />
      </div>
    </div>
  </div>
</div>
`;
